crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2366 @@
+    suite->kem_id = hpke_kem_tab[(rval % nkems + 1)].kem_id;
+
+    /* random kdf */
+    if (RAND_bytes_ex(libctx, &rval, sizeof(rval), OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    suite->kdf_id = hpke_kdf_tab[(rval % nkdfs + 1)].kdf_id;
+
+    /* random aead */
+    if (RAND_bytes_ex(libctx, &rval, sizeof(rval), OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    suite->aead_id = hpke_aead_tab[(rval % naeads + 1)].aead_id;
+    return 1;
+}
+
+/*
+ * @brief return a (possibly) random suite, public key, ciphertext for GREASErs
+ *
+ * @param libctx is the context to use
+ * @param propq is a properties string
+ * @param suite-in specifies the preferred suite or NULL for a random choice
+ * @param suite is the chosen or random suite
+ * @param pub a random value of the appropriate length for sender public value
+ * @param pub_len is the length of pub (buffer size on input)
+ * @param ciphertext buffer with random value of the appropriate length
+ * @param ciphertext_len is the length of cipher
+ * @return 1 for success, otherwise failure
+ */
+static int hpke_good4grease(OSSL_LIB_CTX *libctx, const char *propq,
+                            OSSL_HPKE_SUITE *suite_in,
+                            OSSL_HPKE_SUITE *suite,
+                            unsigned char *pub,
+                            size_t *pub_len,
+                            unsigned char *ciphertext,
+                            size_t ciphertext_len)
+{
+    OSSL_HPKE_SUITE chosen;
+    int crv = 0;
+    int erv = 0;
+    size_t plen = 0;
+    uint16_t kem_ind = 0;
+
+    if (pub == NULL || !pub_len
+        || ciphertext == NULL || !ciphertext_len || suite == NULL)
+        return 0;
+    if (suite_in == NULL) {
+        /* choose a random suite */
+        crv = hpke_random_suite(libctx, propq, &chosen);
+        if (crv != 1)
+            return crv;
+    } else {
+        chosen = *suite_in;
+    }
+    kem_ind = kem_iana2index(chosen.kem_id);
+    if (kem_ind == 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if ((crv = hpke_suite_check(chosen)) != 1)
+        return 0;
+    *suite = chosen;
+    /* publen */
+    plen = hpke_kem_tab[kem_ind].Npk;
+    if (plen > *pub_len)
+        return 0;
+    if (RAND_bytes_ex(libctx, pub, plen, OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    *pub_len = plen;
+    if (RAND_bytes_ex(libctx, ciphertext, ciphertext_len,
+                      OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    return 1;
+err:
+    return erv;
+}
+
+/**
+ * table with identifier and synonym strings
+ * right now, there are 4 synonyms for each - a name, a hex string
+ * a hex string with a leading zero and a decimal string - more
+ * could be added but that seems like enough
+ */
+typedef struct OSSL_HPKE_synonymtab_str {
+    uint16_t id;
+    char *synonyms[4];
+} synonymttab_t;
+
+/**
+ * synonym table for KEMs
+ */
+static synonymttab_t kemstrtab[] = {
+    {OSSL_HPKE_KEM_ID_P256,
```

Comment:
These could actually go into your other tables you have above?

Reply:
Would sorta strongly argue against. I think the new code using those tables is already fairly opaque so further complication for the sake of tidiness seems undesirable unless I'm missing some "win" from doing that.

Reply:
here is a 1-1 mapping between these tables.. Adding an extra value later means making sure you add it to all tables as opposed to 1 table.

Reply:
Fair point on adding new suites later, though I think that'd be better addressed via comments and placement of the tables to make editing more obvious (makes sense to do that latter anyway).

Reply:
did that